### ISSUE TITLE
Enable Start-DbaAgentJob to kick off multiple jobs in parallel and wait for them all to complete

### ISSUE THREAD

**Issue by tonyfountain** [sqlcollaborative/dbatools#6717]
### Summary of new feature

Enable Start-DbaAgentJob to kick off multiple jobs and wait for them all to complete.  Basically allow all jobs to run in parallel and wait for them all to finish (or fail) before returning control to the caller.

### Proposed technical details (if applicable)

Possibly add a -Parallel switch to kick off all the jobs and wait until they are done.

### Latest version of dbatools as of writing

1.0.115

**Comment by tonyfountain** [comment]
FYI - I create a quick snippet to implement my own and seemed pretty simple.  I'll do a pull request tomorrow or Friday with the changes and see if it passes.  I'm thinking of just adding a new switch (something like -Parallel).  The intent is to keep the existing functionality the same, but if -Parallel is $true, then all jobs will be submitted at once, and then will wait for all jobs to complete before returning.  Obviously if -Parallel is $true then both $Wait and $WaitSeconds will also be required.

**Comment by wsmelton** [comment]
We have internal functionality for using runspace.... FYI 😉

**Comment by tonyfountain** [comment]
Thanks @wsmelton ... but my snippet doesn't even require runspaces, or threading... just a simple wait look almost like what is already in the cmdlet when the -Wait parameter is submitted...
```
[System.Object[]]$jobs = Get-DbaAgentJob -SqlInstance $sqlInstance | Where-Object { $PSItem.Name -like "$($jobPrefix)*" };
Start-DbaAgentJob -SqlInstance $sqlInstance -Job $($jobs | % { $PSItem.Name } | Sort-Object) | Out-Null;
While (Get-DbaRunningJob -SqlInstance $sqlInstance -InputObject $jobs) { "Waiting..."; Start-Sleep -Seconds 5; $jobs.Refresh() };
```
Now the first line will be outside the call, as that's what I'm using to get my job list to run... actually line #3 is the key.  Will just need to figure out how to implement it with the existing code and new switch.

**Comment by wsmelton** [comment]
We already do the waiting in the command now to make sure the job started. Are you talking about expanding that or changing how that logic is run overtime?

One challenge I would see with this is dealing with long-running jobs and the connection you have to the instance being killed because of connection timeout. Checking the state of a job every x seconds as well will cause an excessive amount of login traffic potentially as well. I would think that type of functionality is better spent with wrapper scrips of the users; we simply provide the mechanism for starting the job.

**Comment by tonyfountain** [comment]
And as by my snippet, that was pretty simple to be wrap around the existing functionality.  But you're right, this might take a while.  In my context for what I'm doing, it's doing maintenance work but across multiple databases on each job.

**Comment by wsmelton** [comment]
Yeah. Start command should simply start the job. Starting and then waiting for the job to complete is out of scope for this specific command.

Am invoke command might be worthwhile if folks see a need.

**Comment by tonyfountain** [comment]
On the other hand, there is already logic to submit a series of jobs sequentially and wait for all of them to complete.  Adding a -Parallel switch is just a small extension of those options.

**Comment by nvarscar** [comment]
I see this as separating the main loop into two: one for starting the jobs, another one for waiting for them, as simple as that. No runspaces needed.

**Comment by wsmelton** [comment]
And when someone wants to wait on jobs that they want to run sequentially? Do you have a different loop?

**Comment by nvarscar** [comment]
In this case they should run the command in their own loop

**Comment by tonyfountain** [comment]
@wsmelton no, the existing code already takes care of that using the -Wait and -WaitSeconds switches

**Comment by tonyfountain** [comment]
If the concern is adding another loop, give me tomorrow to create a pull request with changes.  Might be easier to explain.  Whether it's accepted or rejected I'll deal with that when we cross that bridge ;).

**Comment by wsmelton** [comment]
Oh, I'm fine with the feature itself. It is just how it will be implemented and the test updated to validate it.

**Comment by tonyfountain** [comment]
Point me to a learning resource (if there is one) for how you do your tests and I'll even dig into that (if that's an option).

**Comment by tonyfountain** [comment]
Ok, I'm adding a switch named RunParallel (since Parallel is now a PS7 keyword!), that when true will be used to start all SQL Agent jobs supplied at once, and wait for them all to finish.  Will be used in conjunction with the -Wait switch.  Question: would it be better to force the caller to pass -Wait when -RunParallel is used, or imply Wait is $true when -RunParallel is used?  I'm thinking the latter myself.

**Comment by nvarscar** [comment]
It would be wrong to imply that user always wants to wait on jobs when -RunParallel is used, even though the behavior is no different from run without such a switch.
I'm still inclined to make the parallel behavior a default one (thus removing the need for a switch and reducing logic complexity). It, however, might be considered a breaking change by some, so I leave it up to maintainers to decide.

**Comment by tonyfountain** [comment]
@nvarscar I'm thinking that way to.  Right now I added a -WaitParallel parameter and tossed in an edit to make -Wait and -WaitParallel mutually exclusive.  Once I got into coding it, that only made better sense.  Although now I'm wondering if I should just combine them into a single parameter.  Instead of a switch have possible values of None, Serialized, Parallel.  None would behave just like not passing -Wait today, Serialized behaves like -Wait does today, Parallel runs all jobs at once and waits for them all to complete (this change).

**Comment by nvarscar** [comment]
Changing the parameter type would be a breaking change. I would suggest to simply re-use the -Wait parameter.

**Comment by tonyfountain** [comment]
I can reuse the -Wait parameter but right now it's a switch.  To give it options to control the different behaviors, that would have to change.

**Comment by nvarscar** [comment]
Why though? 
```powershell
Start-DbaAgentJob -RunParallel -Wait
Start-DbaAgentJob -Wait
Start-DbaAgentJob -RunParallel
Start-DbaAgentJob
```
looks ok to me. A combination of those parameters produces all the behaviours that need to be covered.

**Comment by tonyfountain** [comment]
The way I have it right now, all the options you identified will run except
`Start-DbaAgentJob -RunParallel -Wait`

Oh, and I renamed it -WaitParallel so the calls would look like so:
```
Start-DbaAgentJob -Wait
Start-DbaAgentJob -WaitParallel
Start-DbaAgentJob
```
Basically -WaitParallel ignores -Wait, starts all jobs, and then waits for them all to be complete.

**Comment by nvarscar** [comment]
This is what I'm trying to understand, why introduce a new parameter, when the old one is pretty much doing the same thing?

**Comment by tonyfountain** [comment]
The old parameter -Wait does submits a job, waits for it, submits another job, waits for it.  The new feature will submit all jobs at the same time and wait for them all to complete.  So there are basically two types of waits - serialized and parallel.  I'm thinking it can all be combined into a single parameter (whether it's called -Wait or not is fine - I would prefer -Wait).  But have it more like an enum of what I described above.  I think both behaviors have their merit and use cases.

**Comment by nvarscar** [comment]
Sorry for being unclear. There is no need for two separate types of wait, as this can be achieved by using a single parameter. Pseudocode:
```powershell
foreach ($job in $jobs) { 
  $job.Start()
  if (-Not $RunParallel -and $Wait) { 
    # wait here
  }
}
if ($RunParallel -and $Wait) {
  # wait for each job
}
```

**Comment by tonyfountain** [comment]
Maybe I'm getting hung up on the naming. If I named it -RunParallel or -Parallel, could make it mutually inclusive of -Wait and still get the same behavior.

**Comment by tonyfountain** [comment]
Haha, we cross commented ;)

**Comment by tonyfountain** [comment]
My hang up is this, at a functional level Start-DbaAgentJob is called.  The -Wait parameter tells the cmdlet to wait.  Whether it is serial or parallel doesn't matter at this level, as they're a type of wait.  So all the caller needs to identify, if they want to wait is if they wait each job to be serialized or ran in parallel.  I guess an option I didn't think about, and I think you're eluding to, is what if they want to submit all the jobs in parallel and not wait?  I did think about that, but that's the default behavior now, without any -Wait parameters.  So the parallel option makes no sense here.  One reason why I went from RunParallel to WaitParallel.  This change is not adding functionality to run in parallel, rather its waiting for all the jobs to be completed.  I'm thinking all bases are covered like so:

```
Start-DbaAgentJob
Start-DbaAgentJob -Wait Serialize
Start-DbaAgentJob -Wait Parallel
```

**Comment by nvarscar** [comment]
I agree with this option

> Or we leave -Wait as-is and:
> 
> ```
> Start-DbaAgentJob
> Start-DbaAgentJob -Wait
> Start-DbaAgentJob -Wait -Parallel
> ```

Also, I don't think -Parallel is a keyword and we should be able to use it without any concerns.

**Comment by tonyfountain** [comment]
Yeah, you're right... it's a ForEach-Object switch in PS7.

**Comment by tonyfountain** [comment]
Just looking at the previous change for this cmdlet (https://github.com/sqlcollaborative/dbatools/pull/6674/files)... and I'm not so sure it provided support for multiple servers that did not exist before.  Thoughts?

**Comment by nvarscar** [comment]
Sorry, can you re-phrase your question? I'm not sure what you mean.

**Comment by tonyfountain** [comment]
@nvarscar that pull request was to provide multiple server support.  The way I see it, the prior code already did that, just in two loops instead of one.

**Comment by nvarscar** [comment]
Previous change removed the ability to capture jobs from multiple servers and pipe them into the command - it always expects the -SqlInstance parameter to be present at all times. So yeah I agree, this doesn't look right. Maybe @MikeyBronowski could shed some light?

**Comment by MikeyBronowski** [comment]
> @nvarscar that pull request was to provide multiple server support.  The way I see it, the prior code already did that, just in two loops instead of one.

There was an issue raised for that. It did not work for me either from what I remember.

**Comment by tonyfountain** [comment]
@MikeyBronowski do you remember what sort of behavior you ran into?

**Comment by MikeyBronowski** [comment]
Hope that is clear

```
# created job on two instances A and B, the C instance has no this job

$job = 'MikeyTest'
$instance = 'InstanceA','InstanceC','InstanceB'             # instance without job in the middle of the list
$instance_reverse = 'InstanceB','InstanceC','InstanceA'     # instance without job in the middle of the list
$instance_last_no_job = 'InstanceA','InstanceB','InstanceC' # instance without job at the end of the list
```

```
Get-DbaAgentJob -SqlInstance $instance -Job $job | SELECT SqlInstance, Name | FT
<#
SqlInstance  Name                       
-----------  ----                       
InstanceA    MikeyTest
InstanceB    MikeyTest
#>
```

```
## whatif test
# results for last server from the list (ACB -> B)
Start-DbaAgentJob -SqlInstance $instance -Job $job -WhatIf
# What if: Performing the operation "Starting job MikeyTest" on target "[InstanceB]".

# results for last server from the list (BCA -> A)
Start-DbaAgentJob -SqlInstance $instance_reverse -Job $job -WhatIf
# What if: Performing the operation "Starting job MikeyTest" on target "[InstanceA]".

# no results as the last server from the list has no job (ABC -> null)
Start-DbaAgentJob -SqlInstance $instance_last_no_job -Job $job -WhatIf
# [no output]
```


```
## actual start


# (ACB -> B)
Start-DbaAgentJob -SqlInstance $instance -Job $job
# 1 result -> (ACB -> B)

# history confirms - B only
Get-DbaAgentJobHistory -SqlInstance $instance -Job $job -ExcludeJobSteps| SELECT SqlInstance, Job, RunDate | FT
<# 1 result, but should be 2 (1 for each instance)
SqlInstance  Job    RunDate            
-----------  ---    -------            
InstanceB    MikeyTest 13/08/2020 00:10:42
#>


# (BCA -> A)
Start-DbaAgentJob -SqlInstance $instance_reverse -Job $job
# 1 result -> (BCA -> A)

# history shows two jobs only (each was the last one from the list from each Start-DbaAgentJob call)
Get-DbaAgentJobHistory -SqlInstance $instance -Job $job -ExcludeJobSteps| SELECT SqlInstance, Job, RunDate | FT
<# 2 results, but should be 4 (2 for each instance)
SqlInstance  Job       RunDate            
-----------  ---       -------            
InstanceA    MikeyTest 13/08/2020 00:12:10
InstanceB    MikeyTest 13/08/2020 00:10:42
#>


# (ABC -> null)
Start-DbaAgentJob -SqlInstance $instance_last_no_job -Job $job
# [no output]

# history has not changed, as the last server on the list has no job
Get-DbaAgentJobHistory -SqlInstance $instance -Job $job -ExcludeJobSteps| SELECT SqlInstance, Job, RunDate | FT
<# 2 results, but should be 6 (3 for each instance)
SqlInstance  Job       RunDate            
-----------  ---       -------            
InstanceA    MikeyTest 13/08/2020 00:12:10
InstanceB    MikeyTest 13/08/2020 00:10:42
#>
```

**Comment by tonyfountain** [comment]
@MikeyBronowski thanks for the details!  Makes perfect sense now, I did not consider a situation where I might pass a job across multiple servers where one or more servers did not contain a job.  With that said, I think the root of the issue is this.  In the 2nd and 3rd conditions, $InputObject is set, not appended to.

```
        foreach ($instance in $SqlInstance) {
            try {
                $server = Connect-SqlInstance -SqlInstance $instance -SqlCredential $SqlCredential
            } catch {
                Stop-Function -Message "Error occurred while establishing connection to $instance" -Category ConnectionError -ErrorRecord $_ -Target $instance -Continue
            }

            # Check if all the jobs need to included
            if ($AllJobs) {
                $InputObject += $server.JobServer.Jobs
            }

            # If a specific job needs to be added
            if (-not $AllJobs -and $Job) {
                $InputObject = $server.JobServer.Jobs | Where-Object Name -In $Job
            }

            # If a job needs to be excluded
            if ($ExcludeJob) {
                $InputObject = $InputObject | Where-Object Name -NotIn $ExcludeJob
            }
        }
```
